### ISSUE TITLE
concat_niimgs is not iterator friendly (was: check_niimgs eliminates the first 3D image from an iterator)

### ISSUE THREAD

**Issue by bcipolli** [nilearn/nilearn#463]
```
from nilearn import datasets
from nilearn.image import iter_img, index_img
from nilearn._utils.niimg_conversions import check_niimgs

haxby_dataset = datasets.fetch_haxby_simple()
func_img_1 = index_img(haxby_dataset.func, 0)

# This works
print list(check_niimgs([func_img_1], return_iterator=True))

# This incorrectly returns an empty iterator
print list(check_niimgs(iter_img([func_img_1]), return_iterator=True))

# This throws an error
try:
    print check_niimgs(iter_img([func_img_1]))
except Exception as e:
    print e.__class__
```

**Expected behavior:**
`check_niimgs` called on an iterator of 3D images to return all images.

**Current behavior:**
`check_niimgs` returns a list, or iterator, with the first image missing.  The output from the above:
```
[<nibabel.nifti1.Nifti1Image object at 0x10b3eaf10>]
[]
<type 'exceptions.StopIteration'>
```

**Notes:**
* I'm not sure what use-cases this would hit.  I expected `mean_img` to call `check_niimgs`, but it does not, nor am I clear when an iterator would be passed to `check_niimgs`.
* When `check_niimgs` runs, it grabs the first image to examine image information.  In the case of an iterator, however, it simply returns the remaining iterator (not a new one that contains the first image that's been popped off.
* I'm not sure the best way to fix this.  Issue #462 was an accidental product of me trying to find a solution to this issue.

**Comment by lesteve** [comment]
Looks like a bug to me, a simple fix would be to use itertools.tee to consume the first item but still be able to return the unadulterated iterator. Adding a test would be great too.

**Comment by AlexandreAbraham** [comment]
@bcipolli thanks for spotting this one. By digging in this code, we found out that there is some unconsistent behavior in both `check_niimg` and `check_niimgs`. We want to release soon so we'll fix this one first but we will open a discussion about image loading later.

**Comment by lesteve** [comment]
Hmmm it looks like the check_niimg(s) function are not very iterator friendly in contrary to what the docstring is saying. 

Note this has been the case for a while, and is not specific to the relatively recent return_iterator=True in check_niimgs. To make matters worse, I wasn't able to find tests with check_niimg(s) and iterators.

A couple of examples I just found:

```python
input_iterator = iter([img_3d, img_3d])
img_4d = _utils.check_niimgs(input_iterator)  # Exception in concat_niimgs

input_iterator = iter([img_3d, img_3d])
img_4d = _utils.check_niimg(input_iterator)
img_4d.shape  # (10, 10, 10, 1) instead of (10, 10, 10, 2)
```

The safe way to use check_niimg(s) seems to use lists rather than iterators as arguments and I bet that's the way they are used in our source code.

**Comment by AlexandreAbraham** [comment]
Since we plan to refactor `check_niimg*`, I suggest that we convert iterators into lists and move on until the code is refactored.

**Comment by banilo** [comment]
> Since we plan to refactor check_niimg*

:+1:

**Comment by AlexandreAbraham** [comment]
The first bug evoked by ben is now solved. Here is a new script to test the second one:
```
from nilearn import datasets
from nilearn.image import iter_img, index_img
from nilearn._utils.niimg_conversions import check_niimg_4d

haxby_dataset = datasets.fetch_haxby_simple()
func_img_1 = index_img(haxby_dataset.func, 0)

# This works
print list(check_niimg_4d([func_img_1], return_iterator=True))

# This throws an error
try:
    print check_niimg_4d(iter_img([func_img_1]))
except Exception as e:
    print e.__class__
```

**Comment by GaelVaroquaux** [comment]
What's the status of this guy? Do we have pending work that tries to fix the second bug?

**Comment by AlexandreAbraham** [comment]
#542 fixes the second bug but still has some dirty code and needs review.

**Comment by GaelVaroquaux** [comment]
OK, #542 is fixing 2 critical bugs. So I think that we should try to finish it and merge it before the release.

**Comment by AlexandreAbraham** [comment]
Fixed in #542
